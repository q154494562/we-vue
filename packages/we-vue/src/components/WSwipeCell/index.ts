import WSwipeCell from './WSwipeCell'

export { WSwipeCell }
export default WSwipeCell
