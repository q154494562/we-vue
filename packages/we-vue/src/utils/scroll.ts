type ElementOrWindow = HTMLElement | Window

/**
 * 找到最近的触发滚动事件的元素
 * @param {Element} element
 * @param {Element} rootParent
 * @returns {Element | window}
 */
export function getScrollEventTarget (element: HTMLElement, rootParent: ElementOrWindow = window) {
  let node = element
  while (
    node &&
    node.tagName !== 'HTML' &&
    node.tagName !== 'BODY' &&
    node.nodeType === 1 &&
    node !== rootParent
  ) {
    const { overflowY } = window.getComputedStyle(node)
    if (overflowY === 'scroll' || overflowY === 'auto') {
      return node
    }
    node = <HTMLElement>node.parentNode
  }
  return rootParent
}

// 获取滚动高度
export function getScrollTop (element: ElementOrWindow): number {
  return 'scrollTop' in element ? element.scrollTop : element.pageYOffset
}

// 设置滚动高度
export function setScrollTop (element: ElementOrWindow, value: number) {
  'scrollTop' in element ? (element.scrollTop = value) : element.scrollTo(element.scrollX, value)
}

// 获取元素距离顶部高度
export function getElementTop (element: ElementOrWindow) {
  return (
    (element === window ? 0 : (element as HTMLElement).getBoundingClientRect().top) + getScrollTop(window)
  )
}

export function getVisibleHeight (element: ElementOrWindow) {
  return element === window
    ? element.innerHeight
    : (element as HTMLElement).getBoundingClientRect().height
}
