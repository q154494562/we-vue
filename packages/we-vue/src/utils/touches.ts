function getTouch (event: TouchEvent): Touch {
  return event.changedTouches[0] || event.touches[0]
}

export { getTouch }
