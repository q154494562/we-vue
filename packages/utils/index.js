import Vue from 'vue'
import { getTouch } from './touches'

const isServer = Vue.prototype.$isServer

function isDef (value) {
  return value !== undefined && value !== null
}

function isObj (x) {
  const type = typeof x
  return x !== null && (type === 'object' || type === 'function')
}

export {
  isServer,
  getTouch,
  isDef,
  isObj
}
