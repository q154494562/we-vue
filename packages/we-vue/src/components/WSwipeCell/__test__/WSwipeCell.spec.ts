import Vue, { CreateElement, VNode } from 'vue'
import { mount, shallowMount, Wrapper } from '@vue/test-utils'
import WSwipeCell from '../'
import WCellSwipeButton from '../../WSwipeCellButton'
import { horizontalDrag } from '@/test/unit/utils'

describe('cell-swipe', () => {
  const CellSwipeButtonComponent = Vue.extend({
    render (h: CreateElement): VNode {
      return h(WCellSwipeButton, 'test')
    },
  })

  const getRightBtnsWidth = (wrapper: Wrapper<Vue>) => {
    return (wrapper.vm.$refs.rightBtns as HTMLElement).clientWidth
  }

  beforeEach(() => {
    // to MOCK the clientWidth properfy
    Object.defineProperty(HTMLDivElement.prototype, 'clientWidth', {
      value: 20,
      writable: true,
    })
  })

  afterEach(() => {
    Object.defineProperty(HTMLDivElement.prototype, 'clientWidth', {
      value: 0,
      writable: true,
    })
  })

  test('create', () => {
    const wrapper = shallowMount(WSwipeCell, {
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    expect(wrapper.name()).toBe('w-cell-swipe')
    expect(wrapper.classes()).toContain('weui-cell')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('drag to left and show the buttons', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    horizontalDrag(wrapper.find({ ref: 'cellBd' }), 0, -getRightBtnsWidth(wrapper))

    expect(wrapper.vm.offset).toBe(-getRightBtnsWidth(wrapper))
  })

  test('drag to left with a small distance', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    horizontalDrag(wrapper.find({ ref: 'cellBd' }), 0, -10)

    expect(wrapper.vm.offset).toBe(0)
  })

  test('drag to left with a distance out of range', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    horizontalDrag(wrapper.find({ ref: 'cellBd' }), 0, -(getRightBtnsWidth(wrapper) + 20))

    expect(wrapper.vm.offset).toBe(-getRightBtnsWidth(wrapper))
  })

  test('drag to right and close right buttons', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    // set offset to show the right buttons before drag
    wrapper.setData({
      offset: -getRightBtnsWidth(wrapper),
    })

    horizontalDrag(wrapper.find({ ref: 'cellBd' }), 0, getRightBtnsWidth(wrapper))

    expect(wrapper.vm.offset).toBe(0)
  })

  test('drag to right with a small distance', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    // set offset to show the right buttons before drag
    wrapper.setData({
      offset: -getRightBtnsWidth(wrapper),
    })

    horizontalDrag(wrapper.find({ ref: 'cellBd' }), 0, 10)

    expect(wrapper.vm.offset).toBe(-getRightBtnsWidth(wrapper))
  })

  test('drag to right with a distance out of range', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      slots: {
        right: [CellSwipeButtonComponent, CellSwipeButtonComponent],
      },
    })

    // set offset to show the right buttons before drag
    wrapper.setData({
      offset: -getRightBtnsWidth(wrapper),
    })

    horizontalDrag(wrapper.find({ ref: 'cellBd' }), 0, getRightBtnsWidth(wrapper) + 10)

    expect(wrapper.vm.offset).toBe(0)
  })

  test('click outside', () => {
    const wrapper = mount(WSwipeCell, {
      attachToDocument: true,
      data: function () {
        return {
          offset: 10,
        }
      },
    })

    wrapper.vm.onClickOutside()

    expect(wrapper.vm.offset).toBe(0)
  })
})
