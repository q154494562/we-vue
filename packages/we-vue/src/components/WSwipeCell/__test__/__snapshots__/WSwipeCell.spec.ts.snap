// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cell-swipe create 1`] = `
<div class="weui-cell weui-cell_swiped">
  <div class="weui-cell__bd" style="transform: translate3d(0px, 0px, 0px);">
    <vuecomponent-stub url="function () { [native code] }"><template></template><template></template><template></template></vuecomponent-stub>
  </div>
  <div class="weui-cell__ft">
    <div class="weui-swiped-btn weui-swiped-btn_default">test</div>
    <div class="weui-swiped-btn weui-swiped-btn_default">test</div>
  </div>
</div>
`;
